modules/ximgproc/src/scansegment.cpp

```diff
@@ -0,0 +1,880 @@
+    labelsBuffer = _labelsBuffer.data();
+    clusterBuffer = _clusterBuffer.data();
+    pixelBuffer = _pixelBuffer.data();
+    offsetVec = std::vector<int*>(effectivethreads);
+    int offsetSize = (clusterSize + 1) * sizeof(int);
+    bool offsetAllocated = true;
+    for (int i = 0; i < effectivethreads; i++) {
+        _offsetVec[i] = cv::AutoBuffer<int>(offsetSize);
+        offsetVec[i] = _offsetVec[i].data();
+        if (offsetVec[i] == NULL) {
+            offsetAllocated = false;
+        }
+    }
+    for (int i = 0; i < neighbourCount; i++) {
+        neighbourLocBuffer[i] = (neighbourLoc[i].y * width) + neighbourLoc[i].x;
+    }
+
+    if (labelsBuffer != NULL && clusterBuffer != NULL && pixelBuffer != NULL && offsetAllocated) {
+        setupComplete = true;
+    }
+    else {
+        setupComplete = false;
+
+        if (labelsBuffer == NULL) {
+            CV_Error(Error::StsInternal, "Cannot initialise labels buffer");
+        }
+        if (clusterBuffer == NULL) {
+            CV_Error(Error::StsInternal, "Cannot initialise cluster buffer");
+        }
+        if (pixelBuffer == NULL) {
+            CV_Error(Error::StsInternal, "Cannot initialise pixel buffer");
+        }
+        if (!offsetAllocated) {
+            CV_Error(Error::StsInternal, "Cannot initialise offset buffers");
+        }
+    }
+}
+
+ScanSegmentImpl::~ScanSegmentImpl()
+{
+    // clean up
+    if (!src.empty()) {
+        src.release();
+    }
+    if (!labelsMat.empty()) {
+        labelsMat.release();
+    }
+}
+
+void ScanSegmentImpl::iterate(InputArray img)
+{
+    // ensure setup successfully completed
+    CV_Assert(setupComplete);
+
+    if (img.isMat())
+    {
+        // get Mat
+        src = img.getMat();
+
+        // image should be valid
+        CV_Assert(!src.empty());
+    }
+    else if (img.isMatVector())
+    {
+        std::vector<cv::Mat> vec;
+
+        // get vector Mat
+        img.getMatVector(vec);
+
+        // array should be valid
+        CV_Assert(!vec.empty());
+
+        // merge into Mat
+        cv::merge(vec, src);
+    }
+    else
+        CV_Error(Error::StsInternal, "Invalid InputArray.");
+
+    int depth = src.depth();
+
+    CV_Assert(src.size().width == width && src.size().height == height);
+    CV_Assert(depth == CV_8U);
+    CV_Assert(src.channels() == 3);
+
+    clusterCount = 0;
+    clusterIndex.store(0);
+    clusterID.store(1);
+
+    smallClusters = indexSize / smallClustersDiv;
+
+    // set labels to NONE
+    labelsMat.setTo(NONE);
+
+    // set labels buffer to UNCLASSIFIED
+    std::fill(labelsBuffer, labelsBuffer + indexSize, UNCLASSIFIED);
+
+    // apply light blur
+    cv::medianBlur(src, src, 3);
+
+    // start at the center of the rect, then run through the remainder
+    labBuffer = reinterpret_cast<cv::Vec3b*>(src.data);
+    cv::parallel_for_(cv::Range(0, (int)indexNeighbourVec.size()), PP1(reinterpret_cast<ScanSegmentImpl*>(this)));
```

Comment:
BTW, using of C++11 lambdas with parallel_for looks like:

```
    parallel_for_(Range(0, (int)indexNeighbourVec.size()), [&](const Range& range) {
        for (int i = range.start; i < range.end; i++) {
            OP1(i);
        }
    });
```


OP4 case:

```
    // copy back to labels mat
    parallel_for_(Range(0, (int)indexProcessVec.size()), [&](const Range& range) {
        for (int i = range.start; i < range.end; i++) {
            OP4(indexProcessVec[i]);
        }
    });
```

Reply:
Thanks+++ Appreciate the help. I compiled on my system on C++ 14 and no problems.